rclcpp/include/rclcpp/client.hpp

```diff
@@ -292,34 +419,88 @@ class Client : public ClientBase
     std::shared_ptr<rmw_request_id_t> request_header,
     std::shared_ptr<void> response) override
   {
-    std::unique_lock<std::mutex> lock(pending_requests_mutex_);
-    auto typed_response = std::static_pointer_cast<typename ServiceT::Response>(response);
-    int64_t sequence_number = request_header->sequence_number;
-    // TODO(esteve) this should throw instead since it is not expected to happen in the first place
-    if (this->pending_requests_.count(sequence_number) == 0) {
-      RCUTILS_LOG_ERROR_NAMED(
-        "rclcpp",
-        "Received invalid sequence number. Ignoring...");
+    auto opt = this->get_and_erase_pending_request(request_header->sequence_number);
```

Comment:
nitpick: this is a kind of a "bad" use of `auto` in my opinion, since it's not obvious that the return type here is an `optional`. I would either not use `auto` or use a more descriptive (and ideally not abbreviated) variable name like `optional_pending_request`.

Reply:
Fixed in https://github.com/ros2/rclcpp/pull/1734/commits/14ad9d3966045a7ca308cd5e3fbf1a7be3ad712e.